src/Course/State.hs

```diff
@@ -191,8 +170,33 @@ distinct =
 --
 -- >>> isHappy 44
 -- True
-isHappy ::
-  Integer
-  -> Bool
-isHappy =
-  error "todo: Course.State#isHappy"
+-- isHappy :: Integer -> Bool
+-- isHappy x = computeHapiness x (State (const (0, digits x)))
+
+computeHapiness :: Integer -> State (List Int) Integer -> Bool
+computeHapiness x xs
+  | digitsToNumber (exec xs Nil) == x &&  (eval xs Nil) > 0 = False
+  | (exec xs Nil) == (1 :. Nil) && (eval xs Nil) > 0 = True
+  | otherwise = computeHapiness x squareSumState
+      where
+        squareSumState = xs >>= (\a -> State (\s -> (a + 1, nextState s)))
+        nextState = digits . toInteger . sum . map (join (*))
+
+digits :: Integer -> List Int
+digits = reverse . map P.fromIntegral . unfoldr p
+  where
+    p 0 = Empty
+    p x = Full (x `mod` 10, x `div` 10)
+
+digitsToNumber :: List Int -> Integer
+digitsToNumber = snd . foldRight reducer (1, 0)
```

Comment:
hehe :) that's interesting :) ironically, we could use the state monad here, too, right? we want to keep and update a multiplier as we traverse the list, building up a result. :)

Reply:
yep, that's true, let me refactor that just for fun

Reply:
```
digitsToNumber :: List Int -> Integer
digitsToNumber xs = exec (foldRight reducer (pure 1) xs) 0
  where
    reducer x oldState = oldState >>= (\mult -> State (\acc -> (mult * 10, toInteger x * mult + acc))  )

``` 
it was fun indeed